RadarSDK/RadarBeaconManager.m

```diff
@@ -0,0 +1,217 @@
+//
+//  RadarBeaconManager.m
+//  RadarSDK
+//
+//  Copyright © 2020 Radar Labs, Inc. All rights reserved.
+//
+
+#import "RadarBeaconManager.h"
+
+#import "RadarBeacon+Internal.h"
+#import "RadarDelegateHolder.h"
+#import "RadarLogger.h"
+
+@interface RadarBeaconManager ()
+
+@property (assign, nonatomic) BOOL started;
+@property (nonnull, strong, nonatomic) NSMutableArray<RadarBeaconCompletionHandler> *completionHandlers;
+@property (nonnull, strong, nonatomic) NSMutableSet<NSString *> *nearbyBeaconIdentifers;
+@property (nonnull, strong, nonatomic) NSMutableSet<NSString *> *failedBeaconIdentifiers;
+@property (nonnull, strong, nonatomic) NSArray<RadarBeacon *> *beacons;
+
+@end
+
+@implementation RadarBeaconManager
+
++ (instancetype)sharedInstance {
+    static dispatch_once_t once;
+    static id sharedInstance;
+    if ([NSThread isMainThread]) {
+        dispatch_once(&once, ^{
+            sharedInstance = [self new];
+        });
+    } else {
+        dispatch_sync(dispatch_get_main_queue(), ^{
+            dispatch_once(&once, ^{
+                sharedInstance = [self new];
+            });
+        });
+    }
+    return sharedInstance;
+}
+
+- (instancetype)init {
+    self = [super init];
+    if (self) {
+        _locationManager = [CLLocationManager new];
+        _locationManager.delegate = self;
+
+        _completionHandlers = [NSMutableArray<RadarBeaconCompletionHandler> new];
+
+        _beacons = @[];
+        _nearbyBeaconIdentifers = [NSMutableSet new];
+        _failedBeaconIdentifiers = [NSMutableSet new];
+
+        _permissionsHelper = [RadarPermissionsHelper new];
+    }
+    return self;
+}
+
+- (void)callCompletionHandlersWithStatus:(RadarStatus)status beacons:(NSArray<NSString *> *_Nullable)nearbyBeacons {
+    @synchronized(self) {
```

Comment:
Probably worth adding comment here or somewhere in this class explaining why `synchronized` needed.

Reply:
👍 It's to ensure that we don't try to remove a callback due to a timeout while we're looping through them to call them (same pattern as in `RadarLocationManager`)